src/bootloaders/grub2.c

```diff
@@ -243,18 +247,42 @@ bool grub2_write_kernel(const Grub2Config *config, const Kernel *kernel)
                                          tab);
-                if (config->is_separate) {
-                        cbm_writer_append_printf(config->writer,
-                                                 "echo \"%sinitrd /%s\"\n",
-                                                 tab,
-                                                 kernel->target.initrd_path);
-                } else {
+                cbm_writer_append_printf(config->writer,
+                                         "echo \"%sinitrd %s/%s",
+                                         tab,
+                                         (!config->is_separate)?BOOT_DIRECTORY:"", /* i.e. /boot */
+                                         kernel->target.initrd_path);
+
+                boot_manager_initrd_iterator_init(config->manager, &iter);
+                while (boot_manager_get_initrd_freestanding_entry(&iter, initrd_name)) {
                         cbm_writer_append_printf(config->writer,
-                                                 "echo \"%sinitrd %s/%s\"\n",
-                                                 tab,
-                                                 BOOT_DIRECTORY, /* i.e. /boot */
-                                                 kernel->target.initrd_path);
+                                                 " %s/%s",
+                                                 (!config->is_separate)?BOOT_DIRECTORY:"", /* i.e. /boot */
+                                                 initrd_name);
+                }
+                cbm_writer_append_printf(config->writer, "\"\n");
+        } else {
+                cbm_writer_append_printf(config->writer,
+                                         "echo \"%secho 'Loading initial ramdisk'\"\n",
+                                         tab);
+                boot_manager_initrd_iterator_init(config->manager, &iter);
+                while (boot_manager_get_initrd_freestanding_entry(&iter, initrd_name)) {
+                        if (initrd_counter == 0) {
+                                cbm_writer_append_printf(config->writer,
+                                                         "echo \"%sinitrd %s/%s",
+                                                         tab,
+                                                         (!config->is_separate)?BOOT_DIRECTORY:"", /* i.e. /boot */
+                                                         initrd_name);
+                        } else {
+                                cbm_writer_append_printf(config->writer,
+                                                         " %s/%s",
+                                                         (!config->is_separate)?BOOT_DIRECTORY:"", /* i.e. /boot */
+                                                         initrd_name);
+                        }
+                        initrd_counter++;
                 }
+                if (initrd_counter)
+                        cbm_writer_append_printf(config->writer, "\"\n");
```

Comment:
This entire block needs to be re-written to remove duplication of code. Instead of printing to writer, prepare the entire initrd string and then print prolog and the config string if there are entries. Pseudocode sketch:

```
char *paths = NULL;
if (kernel->initrd_path) {
    paths = string_append(kernel->initrd_path);
}
while (get_next(iter)) {
    paths = string_append(initrd_path);
}
if (strlen(paths)) {
    cbm_writer_append...("echo ... Loading initrd");
    cbm_writer_append...("echo \"%sinitrd ");
    cbm_writer_append...(paths);
    cbm_writer_append...("\"\n");
}
```

Reply:
good improvement

Reply:
I'll do